lang/go/bindings/cne/umem.go

```diff
@@ -0,0 +1,154 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2022 Intel Corporation.
+ */
+
+package cne
+
+/*
+#cgo CFLAGS: -I../../../../usr/local/include/cndp
+#cgo LDFLAGS: -L../../../../usr/local/lib/x86_64-linux-gnu -lcndp -lbsd
+
+#include <cne.h>
+#include <cne_mmap.h>
+#include <pktmbuf.h>
+#include <pktdev.h>
+#include <uds.h>
+#include <uds_connect.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"strconv"
+	"unsafe"
+)
+
+type regionData struct {
+	addr   *C.char
+	pool   *C.pktmbuf_info_t
+	bufCnt uint
+}
+
+// UmemData is the primary structure for UMEM information and data
+type UmemData struct {
+	name  string        // Name of the UMEM
+	mm    *C.mmap_t     // Memory mmap information pointer
+	rInfo []*regionData // Region data information
+}
+
+var umemDataList []*UmemData
+var umemDataMap map[string]*UmemData
+
+func init() {
+	umemDataMap = make(map[string]*UmemData)
+}
+
+// setupUMEM initializes the given data structure for AF_XDP UMEM
+func (sys *System) setupUMEM(umemName string, umem *UmemInfo) error {
+
+	cMmapType := C.CString(umem.MemType)
+	defer C.free(unsafe.Pointer(cMmapType))
+
+	mmapType := (C.mmap_type_t)(C.mmap_type_by_name(cMmapType))
+
+	ud := &UmemData{name: umemName}
+
+	mmapPtr := (*C.mmap_t)(C.mmap_alloc(C.uint(umem.BufCnt), C.uint(umem.BufSize), mmapType))
+	if mmapPtr == nil {
+		return fmt.Errorf("failed to allocate mmap memory %d for umem %s", umem.BufCnt*umem.BufSize, umemName)
+	}
+	ud.mm = mmapPtr
+
+	umemAddr := C.mmap_addr(unsafe.Pointer(ud.mm))
+	ud.rInfo = make([]*regionData, len(umem.Regions))
+
+	for i := 0; i < len(umem.Regions); i++ {
+		ud.rInfo[i] = &regionData{}
+		ri := ud.rInfo[i]
+		ri.bufCnt = umem.Regions[i]
+		ri.addr = (*C.char)(umemAddr)
+		umemAddr = (unsafe.Pointer)((uintptr(umemAddr) + uintptr(ri.bufCnt*umem.BufSize)))
+
+		bufcnt := C.uint(ri.bufCnt)
+		bufsz := C.uint(umem.BufSize)
+		cache := C.uint(sys.jcfg.Defaults().Cache)
+
+		pi := (*C.pktmbuf_info_t)(C.pktmbuf_pool_create(ri.addr, bufcnt, bufsz, cache, nil))
+		if pi == nil {
```

Comment:
if we fail here should we free mmapPtr before returning the error or do we just want to say that we failed for a region in the umem?